### PR TITLE
wgengine/filter: support FilterRules matching on srcIP node caps [capver 100]

### PR DESCRIPTION
**Pull request by bradfitz**
See #12542 for background.

Updates #12542

### CODE REVIEW

File: ipn/ipnlocal/local.go [hunk 1]
```diff
@@ -6839,3 +6839,28 @@ func (b *LocalBackend) startAutoUpdate(logPrefix string) (retErr error) {
...
+	if cap == "" {
+		// Shouldn't happen, but just in case.
+		// But the empty cap also shouldn't be found in Node.CapMap.
+		return false
+	}
+
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	nodeID, ok := b.nodeByAddr[srcIP]
```
Comment (maisem) [inline ipn/ipnlocal/local.go:6857]: `nodeByAddr` is missing docs, can you add them? something like `nodeByAddr maps Tailscale IPs to the respective peers, it doesn't include subnet ips blah blah`

File: wgengine/filter/filter_test.go [hunk 2]
```diff
@@ -40,14 +40,27 @@ const (
 	testDeniedProto  ipproto.Proto = 127 // CRUDP, appropriately cruddy
 )
 
-func m(srcs []netip.Prefix, dsts []NetPortRange, protos ...ipproto.Proto) Match {
-	if protos == nil {
+func m(srcs []netip.Prefix, dsts []NetPortRange, opts ...any) Match {
```
Comment (maisem) [inline wgengine/filter/filter_test.go:43]: use of `any` deserves a comment

Comment (maisem) [reply wgengine/filter/filter_test.go:43]: oh its just a test, still worth a comment

File: ipn/ipnlocal/local.go (hunk 1 above)
Comment (bradfitz) [reply ipn/ipnlocal/local.go:6857]: done

File: wgengine/filter/filtertype/filtertype.go [hunk 3]
```diff
@@ -66,11 +66,23 @@ type CapMatch struct {
...
 type Match struct {
-	IPProto      views.Slice[ipproto.Proto] // required set (no default value at this layer)
-	Srcs         []netip.Prefix
+	// IPProto is the set of IP protocol numbers for which this match applies.
+	// It is required. There is no default value at this layer.
+	// If empty, it doesn't match.
+	IPProto views.Slice[ipproto.Proto]
+
+	// Srcs is the set of source IP prefixes for which this match applies.
+	Srcs []netip.Prefix
```
Comment (maisem) [inline wgengine/filter/filtertype/filtertype.go:75]: say something about how it relates to SrcCaps and SrcContains.

File: wgengine/filter/match.go [hunk 4]
```diff
@@ -32,9 +36,30 @@ func (ms matches) match(q *packet.Parsed) bool {
 	return false
 }
 
-func (ms matches) matchIPsOnly(q *packet.Parsed) bool {
+func srcMatches(m *filtertype.Match, srcAddr netip.Addr, hasCap CapTestFunc) bool {
```
Comment (maisem) [inline wgengine/filter/match.go:39]: doc here too

File: wgengine/filter/filter_test.go (hunk 2 above)
Comment (bradfitz) [reply wgengine/filter/filter_test.go:43]: done

**Review by maisem — APPROVED** [review]

File: wgengine/filter/match.go (hunk 4 above)
Comment (bradfitz) [reply wgengine/filter/match.go:39]: done

File: wgengine/filter/filtertype/filtertype.go (hunk 3 above)
Comment (bradfitz) [reply wgengine/filter/filtertype/filtertype.go:78]: done

**Pull request merged**